File: 240421/겹치지 않게 선분 고르기/select-segments-without-overlap.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    static int N;
    static List<int[]> lines = new ArrayList<>();
    static List<int[]> selected = new ArrayList<>();
    static int result = 0;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        N = scanner.nextInt();

        for (int i = 0; i < N; i++) {
            int x1 = scanner.nextInt();
            int x2 = scanner.nextInt();
            lines.add(new int[]{x1, x2});
        }

        comb(0);
        System.out.println(result);
    }

    static boolean isOverlapped(int[] line1, int[] line2) {
        int x1 = line1[0], x2 = line1[1], x3 = line2[0], x4 = line2[1];
        return (x1 <= x3 && x3 <= x2) || (x1 <= x4 && x4 <= x2) ||
               (x3 <= x1 && x1 <= x4) || (x3 <= x2 && x2 <= x4);
    }

    static boolean isPossible() {
        for (int i = 0; i < selected.size(); i++) {
            for (int j = i + 1; j < selected.size(); j++) {
                if (isOverlapped(selected.get(i), selected.get(j))) {
                    return false;
                }
            }
        }
        return true;
    }

    static void comb(int depth) {
        if (depth == N) {
            if (isPossible()) {
                result = Math.max(result, selected.size());
            }
            return;
        }

        selected.add(lines.get(depth));
        comb(depth + 1);

        selected.remove(selected.size() - 1);
        comb(depth + 1);
    }
}

File: 240421/겹치지 않게 선분 고르기/select-segments-without-overlap.py
N = int(input())
lines = [tuple(map(int, input().split())) for _ in range(N)]

def is_overraped(line1, line2):
    (x1, x2), (x3, x4) = line1, line2
    return (x1 <= x3 <= x2) or (x1 <= x4 <= x2) or (x3 <= x1 <= x4) or (x3 <= x2 <= x4)

def is_possible():
    for i, line1 in enumerate(selected):
        for j, line2 in enumerate(selected):
            if i < j and is_overraped(line1, line2):
                return False

    return True

res = 0
selected = []
def comb(depth):
    global res

    if depth == N:
        if is_possible():
            res = max(res, len(selected))
        return

    selected.append(lines[depth])
    comb(depth + 1)

    selected.pop()
    comb(depth + 1)

comb(0)
print(res)